rust/src/example_iteration.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::io::Read;
+
+use yoke::Yoke;
+
+pub use super::parallel_map::{parallel_map, ParallelMap};
+pub use super::shard_generated::sedpack::io::flatbuffer::shardfile::{root_as_shard, Shard};
+
+pub type Example = Vec<Vec<u8>>;
+type LoadedShard = Yoke<Shard<'static>, Vec<u8>>;
+
+/// Iterate all examples in given shard files.
+pub struct ExampleIterator {
+    example_iterator:
+        std::iter::Flatten<ParallelMap<<Vec<String> as IntoIterator>::IntoIter, ShardProgress>>,
+}
+
+impl ExampleIterator {
+    pub fn new(files: Vec<String>, repeat: bool, threads: usize) -> Self {
+        if repeat {
+            panic!("Not implemented yet: repeat=true");
+        }
+        let example_iterator =
+            parallel_map(get_shard_progress, files.into_iter(), threads).flatten();
+        ExampleIterator { example_iterator }
+    }
+}
+
+impl Iterator for ExampleIterator {
+    type Item = Example;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.example_iterator.next()
+    }
+}
+
+/// Iterator over a single shard file.
+struct ShardProgress {
+    total_examples: usize,
+    used_examples: usize,
+    shard: LoadedShard,
+}
+
+/// Get ShardProgress.
+fn get_shard_progress(file_path: String) -> ShardProgress {
```

Comment:
I see, but the `parallel_map` would then need to deal with lifetimes which could be tricky with threads? Or am I missing something?

Reply:
I did the `std::fs::read` but when I change to `get_shard_progress(file_path: &str)` then I get
```
40 |             parallel_map(get_shard_progress, files.into_iter(), threads).flatten();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `String`, found `&str`
```

I was reading the previous as "paralell map needs the type of the iterator and function pointer to match". Thus I would also need to somehow turn `files` into iterator over `&str`?

Reply:
The first option worked. The second was complaining about receiving a map struct. TBH not sure I see much of a difference here (the iterator is consumed one way or the other no?).

Reply:
I don't know how much difference it makes, but if it's possible to take `&str` instead of `String`, it's better since the caller has more choices. The iterator difference would be to take it as argument instead of `Vec<String>`, again with the idea to give more choices to the user. But if you are the only user, and you know you don't need it, then it's indeed not useful.